### ISSUE TITLE
extraneous versionMetadata.xml file in v1 of 504 moabs on storage root 2

### ISSUE THREAD

**Issue by jmartin-sul** [sul-dlss/preservation_catalog#1393]
**Describe the bug**
It seems that some moabs on storage root 2 (504 as of this issue filing) had an extraneous `versionMetadata.xml` datastream file added to their `v0001` directories, in an attempt to get them to conform with data model changes that post dated the creation of these DOR object versions.  But adding the datastream to the already created moab made it technically invalid:  checksum validation will flag unexpected files, i.e. any files in the moab that aren't explicitly listed in `signatureCatalog.xml`.

**User Impact**
This adds noise to the checksum validation results.  Until this is fixed, it makes it hard to determine at a glance which checksum validation errors are likely due to actual content corruption, if one is just querying high level status or looking at results in aggregate in argo.

**Object(s) In This Error State**

Will post druid list and some detailed examples in comments.

**Expected behavior**

Moabs don't contain files not listed in `signatureCatalog.xml`

**Screenshots**

N/A

#### Detailed explanation

As of 2020-02-21, 504 moabs have status details indicating that the `v0001/data/metadata` dir for the object has a `versionMetadata.xml` file that isn't listed in `signatureCatalog.xml`.  The druid @andrewjbtw mentioned in https://github.com/sul-dlss/preservation_catalog/issues/1324#issuecomment-585956839 fits this profile.  I was worried from what Andrew was saying that a file required by the moab spec was missing, and so I was worried that we'd have to add a file, and maybe re-ship a few hundred objects to archive endpoints.   see also #636, #848.

But I don't think that's the case.  I think that a `versionMetadata.xml` data stream file was put there after the fact, as part of the migration from storage root 1 to storage root 2.  This also seems to match @julianmorley's thinking per a [comment](https://github.com/sul-dlss/preservation_catalog/issues/848#issuecomment-396678061) on #848 ("For disk 2 (although I've only checked 3 of them) the failure mode was all the same: sigCat.xml was not manually created or edited. It was created a few seconds after data/metadata/versionMetadata.xml , but did not contain the checksum for that file.")

Per @blalbrit on #636's description, I think this was done to make the moab "conformant" with a later DOR data model, which had that data stream added later:

> For a subset of objects in the repository, v001 of the object predated the use of the versionMetadata datastream. These 554 objects (all ETDs) do not have a versionMetadata.xml file. These objects are currently showing as errors in the auditWF. Even when later versions have been added to these objects with the expected versionMetadata.xml file, the v001 folder is still failing the audit and, therefore, the objects are failing the audit as well.
>
> These objects are not, in fact, in error - at the time they were deposited to SDR2, they were representative of the model.

But from the moab structure perspective, this is just an unexpected random file that was added after the integrity info for the moab was generated.  This is supported by the following evidence on 8 objects that I spot checked from the 504 results with this error message:
* all `v0001/data/metadata/` files for all spot checked objects were created in late 2011, _except_ for `versionMetadata.xml`, invariably created in early 2013.
* none of the moab `manifests/` files list `versionMetadata.xml` before version 2.
* all the v1 moab manifests were also spit out on the same date as the late add `versionMetadata.xml`.  So while the intent might've been to get the `versionMetadata.xml` files into the moab manifest files, that doesn't seem to have happened.

I'd suggest that we write a script that iterates over all the druids that have that invalid checksum status and that unexpected `versionMetadata.xml` status message, and does the following:
* confirms that the v1 `versionMetadata.xml` file was created well after (over a year?) its sibling fedora datastreams according to the file system time stamps in the relevant moab dir.
* confirms that the **v1** `versionMetadata.xml` file isn't listed in any of the moab `manifests` `vNNNN/signatureCatalog.xml` files.

If that's the case, I'd suggest we delete such v1 `versionMetadata.xml` files.

As a practical matter, I'd suggest that we do this by hand for maybe 10 objects chosen at random from this result set, just to help assure ourselves that we actually understand the problem, and that everything we encounter fits this pattern.

**Comment by jmartin-sul** [comment]
i spot checked 8 objects (all of which turned out to be world readable, thankfully, so i can list druids and details here).

all of them fit the pattern described above.

```ruby
not_ok_complete_moabs_sorted = CompleteMoab.joins(:preserved_object, :moab_storage_root).where.not(status: :ok).order('complete_moabs.status asc, moab_storage_roots.storage_location asc') ; nil

status_details_field = CompleteMoab.arel_table[:status_details]
match_condition = '%Moab file %v0001/data/metadata/versionMetadata.xml was not found in Moab signature catalog%' # more stringent but still matches everything ATM
match_condition = '%Moab file %v0001/data/metadata/versionMetadata.xml was not found in Moab signature catalog%' # more stringent but still matches everything ATM
not_ok_complete_moabs_sorted.where(status_details_field.matches(match_condition)).count # 504 of 513 checksum errors
=> 504

not_ok_complete_moabs_sorted.where(status_details_field.matches(match_condition)).group(:status, :storage_location).pluck(:status, :storage_location, 'count(complete_moabs.id)') # all on storage root 2 so far
=> [["invalid_checksum", "/services-disk02/sdr2objects", 504]]

not_ok_complete_moabs_sorted.where(status_details_field.matches(match_condition)).pluck(:status, :storage_location, :druid, :status_details)
=> [["invalid_checksum",
  "/services-disk02/sdr2objects",
  "bc064vt4021",
  "validate_checksums (actual location: services-disk02; ) Moab file /services-disk02/sdr2objects/bc/064/vt/4021/bc064vt4021/v0001/data/metadata/versionMetadata.xml was not found in Moab signature catalog /services-disk02/sdr2objects/bc/064/vt/4021/bc064vt4021/v0003/manifests/signatureCatalog.xml && CompleteMoab status changed from validity_unknown to invalid_checksum"]
...503 more...]
```

i spot checked 8 other druids:
```sh
[pres@preservation-catalog-prod-02 ~]$ ls -lah /services-disk02/sdr2objects/bc/064/vt/4021/bc064vt4021/v0001/data/metadata/
total 44K
drwxrwxr-x 2 pres pres 4.0K Jan 25  2013 .
drwxrwxr-x 4 pres pres 4.0K Jan 25  2013 ..
-rw-r--r-- 1 pres pres 1.3K Jan 25  2013 contentMetadata.xml
-rw-r--r-- 1 pres pres 4.4K Dec  5  2011 descMetadata.xml
-rw-r--r-- 1 pres pres  607 Dec  5  2011 identityMetadata.xml
-rw-r--r-- 1 pres pres  265 Dec  5  2011 provenanceMetadata.xml
-rw-r--r-- 1 pres pres  284 Dec  5  2011 relationshipMetadata.xml
-rw-r--r-- 1 pres pres  555 Dec  5  2011 rightsMetadata.xml
-rw-r--r-- 1 pres pres 1.1K Dec  5  2011 technicalMetadata.xml
-rw-rw-r-- 1 pres pres  163 Jan 25  2013 versionMetadata.xml
[pres@preservation-catalog-prod-02 ~]$ 
[pres@preservation-catalog-prod-02 ~]$ ls -lah /services-disk02/sdr2objects/bc/064/vt/4021/bc064vt4021/v0003/manifests/
total 40K
drwxrwxr-x 2 pres pres 4.0K Jan 28  2019 .
drwxrwxr-x 4 pres pres 4.0K Jan 28  2019 ..
-rw-rw-r-- 1 pres pres 6.0K Jan 28  2019 fileInventoryDifference.xml
-rw-rw-r-- 1 pres pres 1.6K Jan 28  2019 manifestInventory.xml
-rw-rw-r-- 1 pres pres 7.3K Jan 28  2019 signatureCatalog.xml
-rw-r--r-- 1 pres pres 1.8K Jan 28  2019 versionAdditions.xml
-rw-r--r-- 1 pres pres 4.3K Jan 28  2019 versionInventory.xml
[pres@preservation-catalog-prod-02 ~]$ 
[pres@preservation-catalog-prod-02 ~]$ ls -lah /services-disk02/sdr2objects/bc/064/vt/4021/bc064vt4021/v0001/manifests/
total 28K
drwxrwxr-x 2 pres pres 4.0K Jan 25  2013 .
drwxrwxr-x 4 pres pres 4.0K Jan 25  2013 ..
-rw-rw-r-- 1 pres pres 3.5K Jan 25  2013 fileInventoryDifference.xml
-rw-rw-r-- 1 pres pres 1.6K Jan 25  2013 manifestInventory.xml
-rw-rw-r-- 1 pres pres 2.8K Jan 25  2013 signatureCatalog.xml
-rw-rw-r-- 1 pres pres 3.1K Jan 25  2013 versionAdditions.xml
-rw-rw-r-- 1 pres pres 3.2K Jan 25  2013 versionInventory.xml
[pres@preservation-catalog-prod-02 ~]$ 
[pres@preservation-catalog-prod-02 ~]$ 
[pres@preservation-catalog-prod-02 ~]$ 
[pres@preservation-catalog-prod-02 ~]$ ls -lah /services-disk02/sdr2objects/nz/917/yr/5574/nz917yr5574/v0001/data/metadata/
total 44K
drwxrwxr-x 2 pres pres 4.0K Jan 24  2013 .
drwxrwxr-x 4 pres pres 4.0K Jan 24  2013 ..
-rw-r--r-- 1 pres pres 1.3K Jan 24  2013 contentMetadata.xml
-rw-r--r-- 1 pres pres 6.4K Dec  5  2011 descMetadata.xml
-rw-r--r-- 1 pres pres  629 Dec  5  2011 identityMetadata.xml
-rw-r--r-- 1 pres pres  265 Dec  5  2011 provenanceMetadata.xml
-rw-r--r-- 1 pres pres  271 Dec  5  2011 relationshipMetadata.xml
-rw-r--r-- 1 pres pres  487 Dec  5  2011 rightsMetadata.xml
-rw-r--r-- 1 pres pres 1.2K Dec  5  2011 technicalMetadata.xml
-rw-rw-r-- 1 pres pres  163 Jan 24  2013 versionMetadata.xml
[pres@preservation-catalog-prod-02 ~]$ 
[pres@preservation-catalog-prod-02 ~]$ ls -lah /services-disk02/sdr2objects/nz/917/yr/5574/nz917yr5574/v0003/manifests/
total 36K
drwxrwxr-x 2 pres pres 4.0K Jan 28  2019 .
drwxrwxr-x 4 pres pres 4.0K Jan 28  2019 ..
-rw-rw-r-- 1 pres pres 5.8K Jan 28  2019 fileInventoryDifference.xml
-rw-rw-r-- 1 pres pres 1.6K Jan 28  2019 manifestInventory.xml
-rw-rw-r-- 1 pres pres 7.0K Jan 28  2019 signatureCatalog.xml
-rw-r--r-- 1 pres pres 1.8K Jan 28  2019 versionAdditions.xml
-rw-r--r-- 1 pres pres 4.0K Jan 28  2019 versionInventory.xml
[pres@preservation-catalog-prod-02 ~]$ 
[pres@preservation-catalog-prod-02 ~]$ ls -lah /services-disk02/sdr2objects/nz/917/yr/5574/nz917yr5574/v0001/manifests/
total 28K
drwxrwxr-x 2 pres pres 4.0K Jan 24  2013 .
drwxrwxr-x 4 pres pres 4.0K Jan 24  2013 ..
-rw-rw-r-- 1 pres pres 3.5K Jan 24  2013 fileInventoryDifference.xml
-rw-rw-r-- 1 pres pres 1.6K Jan 24  2013 manifestInventory.xml
-rw-rw-r-- 1 pres pres 2.8K Jan 24  2013 signatureCatalog.xml
-rw-rw-r-- 1 pres pres 3.1K Jan 24  2013 versionAdditions.xml
-rw-rw-r-- 1 pres pres 3.2K Jan 24  2013 versionInventory.xml
[pres@preservation-catalog-prod-02 ~]$ 
[pres@preservation-catalog-prod-02 ~]$ 
[pres@preservation-catalog-prod-02 ~]$ 
...snip, similar deal for sf708wn7291, ck548xg6061, rj088kw6719, zv032kb0761, bg856gy1187, vn684mg0818, bb593hn2771...
```

i also drilled into a couple to see whether v1 `versionMetadata.xml` was listed in any `signatureCatalog.xml`, and nothing, just v2 onward:
```
[pres@preservation-catalog-prod-02 ~]$ grep 'versionMetadata.xml' -R /services-disk02/sdr2objects/bb/593/hn/2771/bb593hn2771/ --include='signatureCatalog.xml' 
/services-disk02/sdr2objects/bb/593/hn/2771/bb593hn2771/v0002/manifests/signatureCatalog.xml:  <entry originalVersion="2" groupId="metadata" storagePath="versionMetadata.xml">
/services-disk02/sdr2objects/bb/593/hn/2771/bb593hn2771/v0003/manifests/signatureCatalog.xml:  <entry originalVersion="2" groupId="metadata" storagePath="versionMetadata.xml">
/services-disk02/sdr2objects/bb/593/hn/2771/bb593hn2771/v0003/manifests/signatureCatalog.xml:  <entry originalVersion="3" groupId="metadata" storagePath="versionMetadata.xml">
/services-disk02/sdr2objects/bb/593/hn/2771/bb593hn2771/v0004/manifests/signatureCatalog.xml:  <entry originalVersion="2" groupId="metadata" storagePath="versionMetadata.xml">
/services-disk02/sdr2objects/bb/593/hn/2771/bb593hn2771/v0004/manifests/signatureCatalog.xml:  <entry originalVersion="3" groupId="metadata" storagePath="versionMetadata.xml">
/services-disk02/sdr2objects/bb/593/hn/2771/bb593hn2771/v0004/manifests/signatureCatalog.xml:  <entry originalVersion="4" groupId="metadata" storagePath="versionMetadata.xml">
/services-disk02/sdr2objects/bb/593/hn/2771/bb593hn2771/v0005/manifests/signatureCatalog.xml:  <entry originalVersion="2" groupId="metadata" storagePath="versionMetadata.xml">
/services-disk02/sdr2objects/bb/593/hn/2771/bb593hn2771/v0005/manifests/signatureCatalog.xml:  <entry originalVersion="3" groupId="metadata" storagePath="versionMetadata.xml">
/services-disk02/sdr2objects/bb/593/hn/2771/bb593hn2771/v0005/manifests/signatureCatalog.xml:  <entry originalVersion="4" groupId="metadata" storagePath="versionMetadata.xml">
/services-disk02/sdr2objects/bb/593/hn/2771/bb593hn2771/v0005/manifests/signatureCatalog.xml:  <entry originalVersion="5" groupId="metadata" storagePath="versionMetadata.xml">
```

same for ck548xg6061.

**Comment by jmartin-sul** [comment]
for all the druids i looked at, it appeared that they had been successfully replicated to all cloud endpoints.  e.g.

```ruby
[159] pry(main)> druid = 'bb593hn2771'
=> "bb593hn2771"
[160] pry(main)> 
[161] pry(main)> ZipPart.joins(zipped_moab_version: [{ complete_moab: [:preserved_object] }, :zip_endpoint]).where(preserved_objects: { druid: druid }).pluck(:druid, 'current_version AS highest_version', 'zipped_moab_versions.version AS zip_version', :endpoint_name, :status)
=> [["bb593hn2771", 5, 5, "aws_s3_east_1", "ok"],
 ["bb593hn2771", 5, 4, "aws_s3_east_1", "ok"],
 ["bb593hn2771", 5, 3, "aws_s3_east_1", "ok"],
 ["bb593hn2771", 5, 2, "aws_s3_east_1", "ok"],
 ["bb593hn2771", 5, 1, "aws_s3_east_1", "ok"],
 ["bb593hn2771", 5, 5, "ibm_us_south", "ok"],
 ["bb593hn2771", 5, 4, "ibm_us_south", "ok"],
 ["bb593hn2771", 5, 3, "ibm_us_south", "ok"],
 ["bb593hn2771", 5, 2, "ibm_us_south", "ok"],
 ["bb593hn2771", 5, 1, "ibm_us_south", "ok"],
 ["bb593hn2771", 5, 5, "aws_s3_west_2", "ok"],
 ["bb593hn2771", 5, 4, "aws_s3_west_2", "ok"],
 ["bb593hn2771", 5, 3, "aws_s3_west_2", "ok"],
 ["bb593hn2771", 5, 2, "aws_s3_west_2", "ok"],
 ["bb593hn2771", 5, 1, "aws_s3_west_2", "ok"]]
```

query from https://github.com/sul-dlss/preservation_catalog/tree/master/db#view-the-zip-parts-for-a-given-druid

presumably they were replicated with the extraneous `versionMetadata.xml`, since that's what was there at the time the archive zips were created?  if we remove those extraneous data streams, do we want to repush new archives?  i think that's mostly a question for @andrewjbtw (though @julianmorley would have to delete the old archive copies first, and so should probably also weigh in on that part).

**Comment by jmartin-sul** [comment]
all the druids that fit this profile at the moment, all on storage root 2:
[prescat1394_druids_with_possible_extraneous_techMD.csv.txt](https://github.com/sul-dlss/preservation_catalog/files/4239844/prescat1394_druids_with_possible_extraneous_techMD.csv.txt)

it's a single column csv (for easier use in scripting), but github "doesn't support that file type", so i renamed the file with an acceptable extension.

**Comment by jmartin-sul** [comment]
this won't really be "in progress" till a meeting i have scheduled for thursday with andrew to work on this.  but self-assigning and moving over so it doesn't get picked up in the meantime, since we've already assigned resources for it.

**Comment by jmartin-sul** [comment]
I paired with @andrewjbtw yesterday afternoon, and we were able to remediate all instances of this extraneous `versionMetadata.xml` problem.  We re-ran checksum validation on the objects in question, and discovered that 2 of them also had the extraneous `technicalMetadata.xml` issue (see #1394), so this ultimately brought 502 moabs into a fully valid state, rather than all 504 that exhibited this problem.

what we did:

* we already had a list of the druids (see the file in https://github.com/sul-dlss/preservation_catalog/issues/1393#issuecomment-589991788 -- sorry, i misnamed that file in retrospect)
* using that druid list, andrew ran:
```sh
while read line ; do echo /services-disk02/sdr2objects/$(~/bin/make-druid-tree.sh $line)/$line/v0001/data/metadata/versionMetadata.xml ; done < prescat1394_druids_with_possible_extraneous_techMD.csv.txt > extraneous-list.txt
```
  * `make-druid-tree.sh` is a shell script that uses gsed to parse a druid into a tree.
```sh
echo -e "$druid" | gsed -r 's/([0-9])([a-zA-Z])/\1\/\2/g; s/([a-zA-Z])([0-9])/\1\/\2/g'
```
  * we were able to do the above easily because of the regularity of the issue:  always the `versionMetadata.xml` file, always in `v0001`, all instances on `/services-disk02`.
* since `extraneous-list.txt` now had a list of full paths (e.g. `/services-disk02/sdr2objects/bb/593/hn/2771/bb593hn2771/v0001/data/metadata/versionMetadata.xml`), we were able to feed the lines in that file to `ls -lh` and page through the results manually to confirm that all of the paths still existed, and that all had the timestamps we expected (late jan 2013).
* then in preservation robots, we made directories for each druid in the folder `/tmp/extraneous-versionMetadata`.
* then another one line shell script that took the list of druids and the list of paths to the extraneous versionMetadata files and moved them to the appropriate folder for each druid:
```sh
while read druid ; do mv -v "$(grep "$druid" extraneous-list.txt)" "$druid" ; done < druids.txt
```
  * before the move, we ran an alternate version of this loop that printed our intended command, to make sure we weren't constructing the wrong thing

Once the extraneous `versionMetadata.xml` had been moved to a temp location (so we wouldn't lose them if something went wrong, or validation failed unexpectedly), i kicked off checksum validation for the remediated objects by looping over the results to the query described [here](https://github.com/sul-dlss/preservation_catalog/issues/1393#issuecomment-589917999), and invoking `Audit::Checksum.validate_druid(druid)` on each one.

All went back to `ok` status except for the two with the [extraneous `technicalMetadata.xml`](#1394) file.

Closing!
